### PR TITLE
Format multi-line braces blocks using do/end

### LINKED ISSUE #4504: Does not Formatter work well on single-line/multi-line block?
(Please excuse my poor English.)
I think there is no similar issues, but I'm sorry if there is duplicate.

```
$ crystal -v
Crystal 0.22.0 (2017-04-20) LLVM 4.0.0
```

-----------------
I think `crystal tool format` does not work well in the following cases on `block`.

### case1: single-line block with newline

`crystal tool format` does not work in the following cases. 
```crystal
a = [1, 2, 3].map { |i|
  i * i
}

# or

a = [1, 2, 3].map do |i|
  i * i
end
```
In both cases above, I expect the following because the block has single-line.
```crystal
a = [1, 2, 3].map { |i| i * i }
```

### case2: multi-line block with braces

In following case,
```crystal
a = [1, 2, 3].map { |i|
  j = i + 1
  j * j
}
```
I think prefer `do...end` over  `{...}` because the block has multi-line.
```crystal
a = [1, 2, 3].map do |i|
  j = i + 1
  j * j
end
```
-----------------

Is there the reason that `crystal tool format` dose not work well in the above cases?

### PR DESCRIPTION
**Pull request by RX14**
Fixes #4504.

### CODE REVIEW

File: src/compiler/crystal/tools/formatter.cr [hunk 1]
```diff
@@ -2480,11 +2480,13 @@ module Crystal
         next_token_skip_space_or_newline
       end
 
-      if @token.keyword?(:do)
+      needs_do_block = node.location.try(&.line_number) != node.end_location.try(&.line_number)
```
Comment (RX14) [inline src/compiler/crystal/tools/formatter.cr]: I'm not entirely sure this is the correct way to detect this but it seems to work.

Comment (asterite) [inline src/compiler/crystal/tools/formatter.cr]: I was going to do it using the same logic, so I think it's OK :-)

**Comment by asterite** [comment]
I'm merging this because this will effectively end discussions about when to use `{ ... }` and `do ... end`. Thank you!!

**Comment by asterite** [comment]
While working on [rufo](https://github.com/asterite/rufo) I realized this is wrong.

For example:

```crystal
foo bar {
  1 + 2
}
```

Here the block is associated with `bar`, but converting it to `do/end` will associate it to `foo`, which changes the semantic of the program. I'm reverting this.

**Comment by RX14** [comment]
@asterite I don't think that `do/end` should have different parsing rules to `{}`, I should be able to switch between the two freely. I'd even suggest going further and *ensuring* that you need braces around calls with do/end/braces so that you'd need to do

```cr
foo(bar) do
  1 + 2
end
```
but either of those suggestions can be implemented without the other.

I think it's the fact this changes semantics which should be fixed, not this PR (but reverting it is good temporarily).

**Comment by RX14** [comment]
I've looked into making a PR for my above suggestion but I'm finding it hard. The crystal parser is very complex and has a lot of accumulated, undocumented state and undocumented abstractions. It's very hard to refactor the parser to take advantage of the added simplicity of having 1 unambiguous parsing rule for blocks.

**Comment by RX14** [comment]
Perhaps @asterite would have fun unifying the two block syntaxes :). I still think it's a bad idea that `do/end` and `{/}` have slightly different semantics, regardless of formatter changes. Ensuring braces around the arguments of functions with blocks seems like a good way to handle this breaking change such that it's always an error when it's an issue.

**Pull request merged**